src/main/java/engineering/swat/watch/impl/JDKRecursiveDirectoryWatcher.java

```diff
@@ -0,0 +1,317 @@
+        }
+        @Override
+        public FileVisitResult preVisitDirectory(Path subdir, BasicFileAttributes attrs) throws IOException {
+            if (!activeWatches.containsKey(subdir)) {
+                isNewDirectory.addLast(true);
+                return super.preVisitDirectory(subdir, attrs);
+            }
+            isNewDirectory.addLast(false);
+            return FileVisitResult.CONTINUE;
+        }
+        @Override
+        public FileVisitResult postVisitDirectory(Path subdir, IOException exc) throws IOException {
+            isNewDirectory.removeLast();
+            return super.postVisitDirectory(subdir, exc);
+        }
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+            if (isNewDirectory.peekLast() == Boolean.TRUE || !seenFiles.contains(file)) {
+                return super.visitFile(file, attrs);
+            }
+            return FileVisitResult.CONTINUE;
+        }
+    }
+
+    private void registerInitialWatches(Path dir) throws IOException {
+        Files.walkFileTree(dir, new InitialDirectoryScan(dir));
+    }
+
+    private List<WatchEvent> registerForNewDirectory(Path dir) {
+        var events = new ArrayList<WatchEvent>();
+        var seenFiles = new HashSet<Path>();
+        var seenDirectories = new HashSet<Path>();
+        try {
+            Files.walkFileTree(dir, new NewDirectoryScan(dir, events, seenFiles, seenDirectories));
+            detectedMissingEntries(dir, events, seenFiles, seenDirectories);
+            return events;
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+
+    private List<WatchEvent> syncAfterOverflow(Path dir) {
+        var events = new ArrayList<WatchEvent>();
+        var seenFiles = new HashSet<Path>();
+        var seenDirectories = new HashSet<Path>();
+        try {
+            Files.walkFileTree(dir, new OverflowSyncScan(dir, events, seenFiles, seenDirectories));
+            detectedMissingEntries(dir, events, seenFiles, seenDirectories);
+            return events;
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    private void detectedMissingEntries(Path dir, ArrayList<WatchEvent> events, HashSet<Path> seenFiles, HashSet<Path> seenDirectories) throws IOException {
+        // why a second round? well there is a race, between iterating the directory (and sending events)
+        // and when the watches are active. so after we know all the new watches have been registered
+        // we do a second scan and make sure to find paths that weren't visible the first time
+        // and emulate events for them (and register new watches)
+        int directoryCount = seenDirectories.size() - 1;
+        while (directoryCount != seenDirectories.size()) {
+            Files.walkFileTree(dir, new OverflowSyncScan(dir, events, seenFiles, seenDirectories));
```

Comment:
Method `detectMissingEntries` is also called from method `registerForNewDirectory`, which uses `NewDirectoryScan`. Is it ok to use `OverflowSyncScan` for the second round in this case?

Reply:
yes, as the 2nd round is the same as overflow.